### PR TITLE
feat(gen5): v26 PPG→HR via normalized ACF peak

### PR DESCRIPTION
**Pull request by Brackyt**
## Status: Draft — blocked on hardware validation

**Do not merge.** Real-hardware validation of `deriveHrFromGen5PpgWaveform` failed against a WHOOP 5 export (fw 50.40.1.0).

### Real-hardware validation failure

| Metric | Result |
|---|---|
| Evaluable rolling windows | 22 (10–12 s each) |
| Measured v18 HR | 90–100 bpm |
| Current ACF-derived HR | 29–160 bpm |

Synthetic sine tests (72/120/45 bpm, noisy, flatline, under-length) pass but are insufficient. No PPG-derived HR should ship until the algorithm is reworked and re-validated on hardware.

### What this PR contains (research)

- `deriveHrFromGen5PpgWaveform` — pure ACF peak picker on concatenated gen5 v26 PPG int16 samples (24 Hz). Abstains on thin/flat/noisy windows; never fabricates BPM.
- Minimum window raised to **240 samples (~10 s @ 24 Hz)** so resting 40–55 bpm can see ≥4 beats.

### Known follow-ups (valid reviewer feedback, deferred)

- Keep window-duration rule consistent with a configurable `sampleHz` parameter.
- Evaluate real ACF neighbors at both BPM boundaries; add monotonic-trend abstention.

## Test plan

- [x] `dart test test/gen5_ppg_hr_test.dart` (synthetic only — insufficient for ship)
- [ ] Re-validate on hardware after algorithm rework
- [ ] Edge #190 remains blocked until this lands with passing hardware validation

### CODE REVIEW

File: lib/src/onehz/foundations/gen5_ppg_hr.dart [hunk 1]
```diff
@@ -0,0 +1,76 @@
+/// Derive BPM from gen5 v26 PPG waveform bursts (24 Hz int16 samples).
+///
+/// Pure, isolate-safe. Absent / degenerate input returns null — never a
+/// fabricated BPM.
+library;
```
[the remark this answers is not in the document]
Comment (Brackyt) [reply lib/src/onehz/foundations/gen5_ppg_hr.dart:5]: Not an analyzer issue: `library;` is a valid unnamed library directive under this repo's SDK constraint (Dart ^3.5). `dart analyze` on `gen5_ppg_hr.dart` reports no issues.

File: lib/src/onehz/foundations/gen5_ppg_hr.dart [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+/// Minimum concatenated PPG length for resting-HR ACF (~10 s @ 24 Hz).
+/// Four 1 s bursts cannot resolve 40–55 bpm (needs ≥4 beats in-window).
+const int kGen5PpgHrMinSamples = 240;
+
+/// Derive BPM from gen5 v26 PPG (24 Hz int16 samples).
+/// [samples] may be one or more concatenated 24-sample bursts.
+/// Returns null if too short, low variance, or no clear ACF peak in 25–230 bpm.
+int? deriveHrFromGen5PpgWaveform(List<int> samples, {double sampleHz = 24.0}) {
+  if (samples.length < kGen5PpgHrMinSamples || sampleHz <= 0) return null;
```
[the remark this answers is not in the document]
Comment (Brackyt) [reply lib/src/onehz/foundations/gen5_ppg_hr.dart:17]: Valid future work — window duration should stay consistent with a configurable `sampleHz` rather than hard-coded 24 Hz assumptions. Deferred; PR is Draft until hardware validation passes.

File: lib/src/onehz/foundations/gen5_ppg_hr.dart [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    xs[i] -= mean;
+    energy += xs[i] * xs[i];
+  }
+  // Population stddev; flatline / near-constant ADC → abstain.
+  final std = math.sqrt(energy / n);
+  if (std < 1.0 || energy <= 0) return null;
+
+  // Lag ↔ BPM: bpm = 60 * hz / lag. Search 25–230 bpm inclusive.
+  final minLag = math.max(1, (60.0 * sampleHz / 230.0).ceil());
+  final maxLag = math.min(n - 1, (60.0 * sampleHz / 25.0).floor());
```
[the remark this answers is not in the document]
Comment (Brackyt) [reply lib/src/onehz/foundations/gen5_ppg_hr.dart:38]: Valid future work — real ACF neighbor evaluation at BPM boundaries and monotonic-trend abstention are needed before ship. Deferred; PR is Draft pending hardware rework (current ACF: 29–160 bpm vs measured 90–100 bpm on 22 evaluable windows).

**Comment by abdulsaheel** [comment]
Reviewed and pushed `51bf6ae`.

## The "never a fabricated BPM" guarantee didn't hold

The header promises *"Absent / degenerate input returns null — never a fabricated BPM."* Measured against this implementation:

```
baseline wander (slow sine drift, no cardiac content) -> 206 bpm
linear ramp (pure DC drift)                           -> 206 bpm
single step edge                                      -> 206 bpm
```

206 bpm is the **top of the 25–230 search range**, returned with full confidence from signals containing no heartbeat at all.

### Root cause

The peak search considered the **boundaries of its own lag range**:

```dart
final left = lag > minLag ? acf[lag - 1] : double.negativeInfinity;
...
if (c <= left || c <= right) continue;
final prominence = c - math.max(left.isFinite ? left : c - 1, ...);
```

At `lag == minLag` the left neighbour is `-infinity`, so the `c <= left` rejection **can never fire**. The prominence fallback then substitutes `c - 1`, making prominence exactly `1.0` — maximal. So the shortest lag was an unconditionally valid "peak", and every monotonically **decaying** autocorrelation got accepted there.

That's the opposite of a rare edge case. A smoothly decaying ACF is precisely what baseline wander, DC drift and motion artifacts produce — so **the quieter and smoother the input, the more confident the fabricated tachycardia**.

### Fix

A periodicity claim needs a real turning point, so only **interior** lags are peak candidates (both neighbours computed, both strictly lower). All three cases above now abstain.

Real signals are unaffected — clean sines still resolve:

| input | before | after |
|---|---|---|
| 45 bpm | 45 | 45 |
| 60 bpm | 60 | 60 |
| 75 bpm | 76 | 76 |
| 100 bpm | 103 | 103 |
| 140 bpm | 144 | 144 |

(The small offsets are integer-lag quantization at 24 Hz, unchanged by this.)

### Why this matters beyond analytics

**edge#190 consumes this** to write a *derived* HR into `decoded_onehz`. A fabricated 206 bpm would have been persisted as a real resting heart rate for any second the strap didn't measure itself — and then fed the whole downstream stack. Worth landing this before that one.

(Separately, on edge#190 I made derived samples unable to evict a measured row, so at least this could never have overwritten real data — but on unmeasured seconds it would have stood as fact.)

## Partly refuting the finding that prompted this

The review note claimed white noise yields a confident BPM **"~25% of the time"**. Measured over 400 trials it's **6/400 = 1.5%**, both before and after this change — noise was never really the problem. The smooth-signal boundary case was, and that one is **100% reproducible**.

## Verification

7 tests added, mutation-verified: restoring the boundary-inclusive search fails exactly the four fabrication tests and nothing else. Full analytics suite **392 passing**.

**Comment by abdulsaheel** [comment]
Reading back through the thread after pushing `51bf6ae` — @coderabbitai had **already found the boundary bug**, and @Brackyt deferred it. Worth closing that loop explicitly:

> *"At 24 Hz, `minLag` is `7`. Therefore, `round(1440 / bestLag)` cannot exceed `206`, despite the documented 230 BPM maximum. The artificial `double.negativeInfinity` neighbor also accepts a nonpeak boundary."*

That's the same defect, and CodeRabbit got there first — including the detail I'd missed, that the documented 230 bpm ceiling is **unreachable** at 24 Hz because `minLag` quantizes to 7. My commit fixes the non-peak-boundary half (interior peaks only, which is what stops the 206 bpm fabrication on wander/ramp/step). The unreachable-230 half is inherent to integer lags at 24 Hz and is really a documentation fix — the honest range is 25–206, not 25–230.

CodeRabbit's other finding — that `240` samples is 10 s **only at 24 Hz**, so a custom `sampleHz` breaks the window rule — is also still valid and still unaddressed. I left it alone given the below.

## The harder question

@Brackyt's hardware validation is the thing that actually decides this PR:

> *ACF: 29–160 bpm vs measured 90–100 bpm on 22 evaluable windows*

**My fix does not close that gap.** It removes a fabrication mode (confident 206 bpm from signals with no cardiac content) and it keeps clean sines resolving correctly, but a 29–160 spread against a true 90–100 is a signal-processing problem several layers deeper than peak-boundary handling — most likely bandpass/detrending before the ACF, plus the burst-adjacency issue CodeRabbit flagged on edge#190 (the buffer will concatenate captures with missing intermediate bursts, which would smear the autocorrelation exactly like this).

Combined with @Brackyt's note on edge#190 that the whole v26 PPG line was built on a premise that turned out to be wrong — the strap *does* give usable v18 data — my recommendation is to **close this along with edge#190** rather than keep polishing it. A derived bpm with no RR is strictly worse than measured v18 HR + RR, and it costs real complexity to get right.

If it's kept, the boundary fix is worth having regardless, since the function would otherwise fabricate on any smooth artifact.

Your call — I'm happy either way, and I'd rather flag it than quietly keep improving something that shouldn't ship.

**Comment by Brackyt** [comment]
Same conclusion, closing this
